polaris-migrator/src/migrations/replace-sass-z-index/replace-sass-z-index.ts

```diff
@@ -0,0 +1,119 @@
+import type {API, FileInfo, Options} from 'jscodeshift';
+import postcss, {Plugin} from 'postcss';
+import valueParser, {FunctionNode, Node} from 'postcss-value-parser';
+
+import {POLARIS_MIGRATOR_COMMENT} from '../../constants';
+
+interface PluginOptions extends Options {
+  namespace?: string;
+}
+
+const processed = Symbol('processed');
+
+function isNumericOperator(node: Node): boolean {
+  return (
+    node.value === '+' ||
+    node.value === '-' ||
+    node.value === '*' ||
+    node.value === '/' ||
+    node.value === '%'
+  );
+}
+
+const zIndexMap = {
+  content: '--p-z-1',
+  overlay: '--p-z-2',
+};
+
+const isValidElement = (
+  element: unknown,
+): element is keyof typeof zIndexMap => {
+  return Object.keys(zIndexMap).includes(element as string);
+};
+
+const hasMoreThanOneArgument = (node: FunctionNode) => {
+  if (!node.nodes || !Array.isArray(node.nodes)) return false;
+  return node.nodes && node.nodes.length > 1;
+};
+
+const plugin = (options: PluginOptions = {}): Plugin => {
+  const namespace = options?.namespace || '';
+  // This migration could be run over sass with
+  // proper legacy-polaris namespacing
+  // or some other pre-processed / post-processed css
+  // in which case we'll still want to transform a non-namespaced z-index fn.
+  const functionName = namespace ? `${namespace}.z-index` : 'z-index';
+  const isZIndexFn = (node: Node): node is FunctionNode => {
+    return node.type === 'function' && node.value === functionName;
+  };
+  return {
+    postcssPlugin: 'replace-sass-z-index',
+    Declaration(decl) {
+      // @ts-expect-error - Skip if processed so we don't process it again
+      if (decl[processed]) return;
+
+      const parsed = valueParser(decl.value);
+
+      let containsZIndexFn = false;
+      let containsCalculation = false;
+      let containsSecondArgument = false;
+
+      parsed.walk((node) => {
+        if (isZIndexFn(node)) containsZIndexFn = true;
+        if (isNumericOperator(node)) containsCalculation = true;
+
+        if (!isZIndexFn(node)) return;
+
+        if (hasMoreThanOneArgument(node)) {
+          // If there's more than one argument to the zIndex fn
+          // We assume they're passing in a custom map
+          // In this case its unlikely this will resolve to a polaris token value
+          // transform legacy zIndex usage to map-get and move on.
+          containsSecondArgument = true;
+          node.value = 'map-get';
+        } else {
+          const element = node.nodes[0].value ?? '';
```

Comment:
```suggestion
          const element = node.nodes[0]?.value ?? '';
```

The prior validation only guarantees there isn't more than one arg, not that there is args. We should be able to get this flagged by TypeScript ([with `noUncheckedIndexAccess` enabled](https://www.typescriptlang.org/play?noUncheckedIndexedAccess=true#code/MYewdgzgLgBGIBMCmEBcMDeMBuBDANgK5LrQBOAlmAOYwC+A2gLowC8MzAsAFA8D0fGAHUKUABYwABvACqYYGKTAA1kgQBJMMgAeagILBgKCJJhIwuAEb41PUJFh4iSNnEQoGABiYA6J8R4gA))